apps/emqx_retainer/src/emqx_retainer.erl

```diff
@@ -74,11 +75,12 @@ on_session_subscribed(_, Topic, #{rh := Rh, is_new := IsNew}) ->
 
 %% @private
 dispatch(Pid, Topic) ->
-    Msgs = case emqx_topic:wildcard(Topic) of
-               false -> read_messages(Topic);
-               true  -> match_messages(Topic)
-           end,
+    MsgsT = case emqx_topic:wildcard(Topic) of
+                false -> read_messages(Topic);
+                true  -> match_messages(Topic)
+            end,
     Now = erlang:system_time(millisecond),
+    Msgs = drop_banned_messages(MsgsT),
```

Comment:
Q: should we also delete messages from banned clients from storage at this point? :thinking: